conda/core/solve.py

```diff
@@ -1029,6 +1039,251 @@ def _prepare(self, prepared_specs):
+        pool = state["pool"]
+
+        # Set different solver options
+        solver_args = (prefix_data,) if context.force_reinstall else ()
+        solver_options = [(api.SOLVER_FLAG_ALLOW_DOWNGRADE, 1)]
+        if context.channel_priority is ChannelPriority.STRICT:
+            solver_options.append((api.SOLVER_FLAG_STRICT_REPO_PRIORITY, 1))
+        solver_postsolve_flags = [
+            (api.MAMBA_NO_DEPS, context.deps_modifier == DepsModifier.NO_DEPS),
+            (api.MAMBA_ONLY_DEPS, context.deps_modifier == DepsModifier.ONLY_DEPS),
+            (api.MAMBA_FORCE_REINSTALL, context.force_reinstall),
+        ]
+
+        solver = api.Solver(pool, solver_options, *solver_args)
+        solver.set_postsolve_flags(solver_postsolve_flags)
+
+        # Configure jobs
+
+        # 1. Lock currently installed ones, if requested
+        if context.update_modifier is UpdateModifier.FREEZE_INSTALLED:
+            solver.add_jobs([p for p in prefix_data.package_records], api.SOLVER_LOCK)
+
+        # 2. Install/update user requested packages
+        specs_to_add = self._collect_specs_to_add(state)
+        solver_task = api.SOLVER_UPDATE if self._command == "update" else api.SOLVER_INSTALL
+        solver.add_jobs(specs_to_add, solver_task)
+
+        # 3. Special cases
+        # 3a. Handle aggressive updates
+        if not context.force_reinstall:
+            installed_names = [rec.name for rec in state["installed_pkgs"]]
+            always_update = [pkg.name for pkg in context.aggressive_update_packages
+                             if pkg.name in installed_names]
+            solver.add_jobs(always_update, api.SOLVER_UPDATE)
+        # 3b. Pin constrained packages in env
+        python_pin = self._pin_python(state)
+        if python_pin:
+            solver.add_pin(python_pin)
+        for pin in self._pinned_specs():
+            solver.add_pin(pin)
+
+        state["solver"] = solver
+        return solver
+
+    def _configure_solver_for_remove(self, state):
+        from mamba import mamba_api as api
+
+        solver_options = [
+            (api.SOLVER_FLAG_ALLOW_DOWNGRADE, 1),
+            (api.SOLVER_FLAG_ALLOW_UNINSTALL, 1)
+        ]
+        if context.channel_priority is ChannelPriority.STRICT:
+            solver_options.append((api.SOLVER_FLAG_STRICT_REPO_PRIORITY, 1))
+        solver = api.Solver(state["pool"], solver_options)
+
+        history = api.History(self.prefix)
+        history_map = history.get_requested_specs_map()
+        solver.add_jobs(
+            [s.conda_build_form() for s in history_map.values()],
+            api.SOLVER_USERINSTALLED,
+        )
+        specs = [s.conda_build_form() for s in self.specs_to_remove]
+        solver.add_jobs(specs, api.SOLVER_ERASE | api.SOLVER_CLEANDEPS)
+
+        state["solver"] = solver
+        return solver
+
+    def _collect_specs_to_add(self, state):
+        """
+        Reimplement the logic found in super()._add_specs(), but simplified.
```

Comment:
If this is a simplification, let's see if we can refactor _add_specs to simplify as well.